torch/csrc/jit/python/pybind_utils.h

```diff
@@ -289,6 +289,16 @@ inline InferredType tryToInferType(py::handle input) {
       TORCH_INTERNAL_ASSERT(class_type);
       return InferredType(class_type);
     }
+
+    auto isModule = py::hasattr(input, "_concrete_type");
```

Comment:
Can you refactor this to work with normal nn modules as well ?

Reply:
How can we do that? The JIT type for a normal `nn.Module` doesn't exist at this point in time when the concrete type for its parent is still being constructed.

Reply:
We should compile it if it's not scripted already

Reply:
I tried something like this:
```
    auto module_type = py::module::import("torch").attr("nn").attr("Module");
    if (py::isinstance(input, module_type)) {
      auto scripted_module = py::module::import("torch").attr("jit").attr("script")(input);
      auto concrete_type = py::getattr(scripted_module, "_concrete_type");
      auto jit_type = py::getattr(concrete_type, "jit_type");
      auto type = py::cast<ClassTypePtr>(jit_type);
      return InferredType(type);
    }
```

And ran into some weird error related to double compilation of a `Module`:
```
======================================================================
ERROR: test_submodule_tuple_attribute (__main__.TestScriptPy3)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_jit_py3.py", line 521, in test_submodule_tuple_attribute
    scripted_m = torch.jit.script(MyModel())
  File "/data/users/meghanl/repos/pytorch/torch/jit/__init__.py", line 1355, in script
    return torch.jit._recursive.create_script_module(obj, torch.jit._recursive.infer_methods_to_compile)
  File "/data/users/meghanl/repos/pytorch/torch/jit/_recursive.py", line 325, in create_script_module
    return create_script_module_impl(nn_module, concrete_type, stubs_fn)
  File "/data/users/meghanl/repos/pytorch/torch/jit/_recursive.py", line 379, in create_script_module_impl
    script_module = torch.jit.RecursiveScriptModule._construct(cpp_module, init_fn)
  File "/data/users/meghanl/repos/pytorch/torch/jit/__init__.py", line 1736, in _construct
    init_fn(script_module)
  File "/data/users/meghanl/repos/pytorch/torch/jit/_recursive.py", line 360, in init_fn
    scripted = create_script_module_impl(orig_value, sub_concrete_type, infer_methods_to_compile)
  File "/data/users/meghanl/repos/pytorch/torch/jit/_recursive.py", line 383, in create_script_module_impl
    create_methods_from_stubs(concrete_type, stubs)
  File "/data/users/meghanl/repos/pytorch/torch/jit/_recursive.py", line 299, in create_methods_from_stubs
    concrete_type._create_methods(defs, rcbs, defaults)
RuntimeError: Can't redefine method: forward on class: __torch__.torch.nn.modules.linear.Linear
```

Something like this works, but I don't think it's the right approach:
```
def infer_concrete_type_builder(nn_module):
    ...
    ...
        if isinstance(value, collections.abc.Iterable):
            converted = []
            for t in value:
                if isinstance(t, torch.nn.Module):
                    converted.append(torch.jit.script(t))
            value = type(value)(converted)
        attr_type = infer_type(name, value)
```

Reply:
Any suggestions on how to proceed? :(

Reply:
Yea, the below LGTM. 
```
    auto module_type = py::module::import("torch").attr("nn").attr("Module");
    if (py::isinstance(input, module_type)) {
      auto scripted_module = py::module::import("torch").attr("jit").attr("script")(input);
      auto concrete_type = py::getattr(scripted_module, "_concrete_type");
      auto jit_type = py::getattr(concrete_type, "jit_type");
      auto type = py::cast<ClassTypePtr>(jit_type);
      return InferredType(type);
    }
```

we should just figure out what the weirrd bug is

Reply:
Okay, finally got back to this. Due to the existence of a global type store, the concrete type for the module gets reused between this extra compilation that I added in the snippet above (during type construction for the parent module) and another that occurs later because the module is also a member of `self._modules` for the parent module.

One way to fix this is to cache the compilation results for modules and reuse those as well. I've done this in the latest version that I just uploaded (in a crude way). If this makes sense, I can clean it up a bit and make it more like the type store we have now.

Reply:
> and another that occurs later because the module is also a member of self._modules

Would it be possible just to check reuse the type when it gets used in _modules ? cc @suo is the right person to chime in here.